pkg/policy/mapstate.go

```diff
@@ -137,24 +137,36 @@ func (keys MapState) RedirectPreferredInsert(key Key, entry MapStateEntry) {
 // from the localhost. It inserts the Key corresponding to the localhost in
 // the desiredPolicyKeys if the localhost is allowed to communicate with the
 // endpoint.
-func (keys MapState) DetermineAllowLocalhostIngress(l4Policy *L4Policy) {
+func (keys MapState) DetermineAllowLocalhostIngress(isDeny bool) {
 	if option.Config.AlwaysAllowLocalhost() {
 		derivedFrom := labels.LabelArrayList{
 			labels.LabelArray{
 				labels.NewLabel(LabelKeyPolicyDerivedFrom, LabelAllowLocalHostIngress, labels.LabelSourceReserved),
 			},
 		}
-		keys[localHostKey] = NewMapStateEntry(derivedFrom, false)
-		if !option.Config.EnableRemoteNodeIdentity {
-			keys[localRemoteNodeKey] = NewMapStateEntry(derivedFrom, false)
+		if !isDeny {
+			keys[localHostKey] = NewMapStateEntry(derivedFrom, false)
+			if !option.Config.EnableRemoteNodeIdentity {
+				keys[localRemoteNodeKey] = NewMapStateEntry(derivedFrom, false)
+			}
+		} else {
+			// For the policy deny we will delete it from the map.
+			delete(keys, localHostKey)
+			if !option.Config.EnableRemoteNodeIdentity {
+				delete(keys, localRemoteNodeKey)
+			}
```

Comment:
AFAIU, that will only remove L3-only rules for traffic from the local host. What if there is a L4 or L3+L4 deny rule?

By the way, how can someone add a L4-only rule without blocking access from the local host...?

Reply:
>AFAIU, that will only remove L3-only rules for traffic from the local host. What if there is a L4 or L3+L4 deny rule?

How is that different than the way we currently doing it for the non-deny policies?

Reply:
With allow policies, adding a new L3+L4 rule for traffic from the host won't overwrite the existing allow-all-from-host rule:
```
$ tail -n8 test.yaml 
  ingress:
  - fromEntities:
    - host
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP 
$ kubectl create -f test.yaml
$ sudo cilium bpf policy get 2880 | grep Ingress
Ingress     reserved:host                                      80/TCP       NONE         0        0         
Ingress     reserved:host                                      ANY          NONE         13206    93
```
I don't think users can break ingress from host to pod with allow policies. But it looks like they can with deny policies, as follows (not tested):
```yaml
  ingressDeny:
  - fromEntities:
    - host
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP 
```